### PR TITLE
Code review

### PR DESCRIPTION
**Pull request by PerryBaran**

### CODE REVIEW

File: src/Itinerary.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+class Itinerary {
+  constructor(ports) {
+    this.ports = ports;
+  }
+}
+
+module.exports = Itinerary;
```
Comment (PerryBaran) [inline src/Itinerary.js:7]: looks good

File: src/Port.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    this.ships = [];
+  }
+
+  addShip(ship) {
+    this.ships.push(ship);
+  }
+
+  removeShip(ship) {
+    this.ships.pop(ship);
+  }
```
Comment (PerryBaran) [inline src/Port.js:13]: This isn't how `pop` works. `pop` removes the last element from the array, and doesn't take any arguments. To remove an item from a specific place from an array you need to use `indexOf` to find the index of the ship in the array and `splice` to remove the ship at that index, i.e.
```
removeShip(ship) {
  const shipIndex = this.ships.indexOf(ship);
  this.ships.splice(shipIndex, 1);
};
```

File: src/Ship.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    this.previousPort = this.currentPort;
+    this.currentPort = null;
+  }
+  dock() {
+    const itinerary = this.itinerary;
+    const previousPotIndex = itinerary.ports.indexOf(this.previousPort);
+
+    this.currentPort = itinerary.ports[previousPotIndex + 1];
+    this.currentPort.addShip(this);
+  }
```
Comment (PerryBaran) [inline src/Ship.js:26]: remember to add an empty line between you methods to improve readability, i.e.
```
constructor() {
  ... 
}

setSail() {
  ...
}

dock() {
  ...
}
```

File: src/Ship.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    const currentPortIndex = itinerary.ports.indexOf(this.currentPort);
+
+    this.currentPort.removeShip(this);
+
+    if (currentPortIndex === itinerary.ports.length - 1) {
+      throw new Error("End of itinerary reached");
+    }
+    this.previousPort = this.currentPort;
+    this.currentPort = null;
+  }
```
Comment (PerryBaran) [inline src/Ship.js:19]: Instead of assigning the `itinerary` to a variable, you could assign the `ports` to the variable, as you only ever use the itinerary to access the ports, i.e.
```
const ports = this.itinerary.ports;
const currentPortIndex = ports.indexOf(this.currentPort);

this.currentPort.removeShip(this);

if (currentPortIndex === ports.length - 1) {
  throw new Error("End of itinerary reached");
}

this.previousPort = this.currentPort;
this.currentPort = null;
```

File: src/Ship.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  dock() {
+    const itinerary = this.itinerary;
+    const previousPotIndex = itinerary.ports.indexOf(this.previousPort);
+
+    this.currentPort = itinerary.ports[previousPotIndex + 1];
+    this.currentPort.addShip(this);
```
Comment (PerryBaran) [inline src/Ship.js:25]: As above, could assign the `ports` to a variable instead of `itinerary` as you only ever use the `itinerary` to access the `ports`

File: __tests__/Itinerary.test.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    expect(new Itinerary()).toBeInstanceOf(Object);
+  });
+
+  it("can have ports", () => {
+    const dover = new Port("Dover");
+    const calais = new Port("Calais");
+
+    const itinerary = new Itinerary([dover, calais]);
+
+    expect(itinerary.ports).toEqual([dover, calais]);
```
Comment (PerryBaran) [inline __tests__/Itinerary.test.js:17]: Should be using stubs/mocks to replace the Port. As these tests should only relate to the functionality of `Itinerary` and shouldn't depend on `Port`. e.g.
```
const dover = jest.fn();
const calais = jest.fn();

const itinerary = new Itinerary([dover, calais]);

expect(itinerary.ports).toEqual([dover, calais]);
```
In this example I assigned the ports to `jest.fn()`, but really you could assign these ports to any value, as the only thing this test actually checks is that passing an argument to Itinerary assigns that argument to `itinerary.ports`

File: __tests__/Port.test.js [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+      queenMary = jest.fn();
+    });
+
+    it("can be instantiated", () => {
+      expect(new Port()).toBeInstanceOf(Object);
+    });
+
+    it("port has a name", () => {
+      expect(port.name).toBeDefined();
+    });
```
Comment (PerryBaran) [inline __tests__/Port.test.js:23]: I would add an expect statement to check that `port.name` has the expected value, e.g.
```
expect(port.name).toBe("name");
```

File: __tests__/Port.test.js [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+    beforeEach(() => {
+      port = new Port("name");
+      ship = jest.fn();
+      titanic = jest.fn();
+      queenMary = jest.fn();
+    });
+
+    it("can be instantiated", () => {
+      expect(new Port()).toBeInstanceOf(Object);
```
Comment (PerryBaran) [inline __tests__/Port.test.js:18]: instead of calling `new Port()` here, you could reference the `port` you have already defined i.e.
```
expect(port).toBeInstanceOf(Object);
```

File: __tests__/Port.test.js [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+      expect(port.ships).toContain(ship);
+    });
+
+    it("can remove a ship", () => {
+      port.addShip(titanic);
+      port.addShip(queenMary);
+      port.removeShip(queenMary);
+
+      expect(port.ships).toEqual([titanic]);
+    });
```
Comment (PerryBaran) [inline __tests__/Port.test.js:37]: In both of these tests you have 3 different ships defined. You could edit the first test to make use of one of the ships used in your second test, removing the need for the `ship` variable, i.e.
```
port.addShip(titanic);

expect(port.ships).toContain(titanic);
```

File: __tests__/Ship.test.js [hunk 10]
```diff
@@ -0,0 +1,68 @@
+const Ship = require("../src/Ship");
+const Port = require("../src/Port");
+const Itinerary = require("../src/Itinerary");
```
Comment (PerryBaran) [inline __tests__/Ship.test.js:3]: These imports aren't used and should be deleted

File: __tests__/Ship.test.js [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+      expect(ship).toBeInstanceOf(Object);
+    });
+
+    it("gets added to port on instantiation", () => {
+      expect(dover.addShip).toHaveBeenCalledWith(ship);
+    });
+
+    it("has a starting port", () => {
+      expect(ship.currentPort).toBe(dover);
+    });
```
Comment (PerryBaran) [inline __tests__/Ship.test.js:44]: Should add tests for all instantiated values, i.e. add a test to check the `previousPort` is instantiated as `null`

File: __tests__/Ship.test.js [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+      expect(ship.currentPort).toBe(calais);
+      expect(calais.addShip).toHaveBeenCalledWith(ship);
+    });
+
+    it("can set sail", () => {
+      ship.setSail();
+
+      expect(ship.currentPort).toBeFalsy();
+      expect(dover.removeShip).toHaveBeenCalledWith(ship);
+    });
```
Comment (PerryBaran) [inline __tests__/Ship.test.js:59]: I would swap the order of these tests to improve readability. It makes sense to check `setSail` before `dock` as you have to call `setSail` in the `dock` test

Comment (PerryBaran) [inline __tests__/Ship.test.js:59]: I would also add an expect statement to check that `previousPort` was set to the previous `currentPort` in this test

File: __tests__/Port.test.js (hunk 9 above)
Comment (PerryBaran) [inline __tests__/Port.test.js:37]: This test passes despite your broken `removeShip` functionality because you remove the last ship you added. If you were you to change your test to:
```
port.addShip(titanic);
port.addShip(queenMary);
port.removeShip(titanic);

expect(port.ships).toEqual([queensMary]);
```
you'll see that the test fails

**Review by PerryBaran — COMMENTED** [review]
Overall really good, code is well formatted. Logic is good for the most part, although your `removeShip` functionality doesn't work correctly and your test wasn't written in a way to catch this problem.

You should remember to add a README to your projects. These are important as they give act as documentation for your project, describing: what this project is, the most important parts of the project and how to use your project. In my README's I like to include:
- a brief description about what this project is, and why I made it
- a table of contents
- list of dependencies (atm this is just jest, you can find these dependencies in your package.json)
- how to setup the project locally
- available scripts (again, in your package.json) e.g. `npm test` runs all tests
- Could add a section about what you learnt while making this project and what you would do differently if you had more time
- attribution

You can also look into markdown syntax, which will allow you format your README. [you can read more about that here](https://www.markdownguide.org/basic-syntax/)

Also, one last thing. Make sure to add a `.gitignore` before your first commit. Although this won't often come up, the way we set up the code review reverts the main branch to the initial commit and as your initial commit contained `node_modules` I wasn't able to get straight into the code review and had to find a work around to get it working (it won't show all the files changed if there are too many)

File: src/Port.js (hunk 2 above)
Comment (Chris-ONeill45) [reply src/Port.js:13]: Amended

File: src/Ship.js (hunk 3 above)
Comment (Chris-ONeill45) [reply src/Ship.js:26]: Corrected

File: __tests__/Ship.test.js (hunk 10 above)
Comment (Chris-ONeill45) [reply __tests__/Ship.test.js:3]: Deleted imports

File: src/Ship.js (hunk 4 above)
Comment (Chris-ONeill45) [reply src/Ship.js:19]: Amended, tests still passing.

File: __tests__/Itinerary.test.js (hunk 6 above)
Comment (Chris-ONeill45) [reply __tests__/Itinerary.test.js:17]: I though I did these on the Itinerary.test.js file, perhaps I didn't save it and went back. Thanks for spotting this.

File: src/Ship.js (hunk 5 above)
Comment (Chris-ONeill45) [reply src/Ship.js:25]: Amended like the one above. Test's still passing.

File: __tests__/Port.test.js (hunk 7 above)
Comment (Chris-ONeill45) [reply __tests__/Port.test.js:23]: Your amendment was how i originally wrote the code. However, during the 'Isolating with Spies' section of the track, I amended it to use a mock function. I guess i didn't need to do that, as best practice was how it was originally written?

File: __tests__/Ship.test.js (hunk 12 above)
Comment (Chris-ONeill45) [reply __tests__/Ship.test.js:59]: Hi Perry, to do this would I add a const and then write the expect? Like below?

 it("can set sail", () => {
      **const previousPort = ship.currentPort;**

      ship.setSail();

      expect(ship.currentPort).toBeFalsy();
      expect(dover.removeShip).toHaveBeenCalledWith(ship);
      **expect(ship.previousPort).toBe(previousPort);**

Comment (Chris-ONeill45) [reply __tests__/Ship.test.js:59]: Hi Perry. To complete this would i add a const then the expect? Like below?

 it("can set sail", () => {
      const previousPort = ship.currentPort;
      ship.setSail();

      expect(ship.currentPort).toBeFalsy();
      expect(dover.removeShip).toHaveBeenCalledWith(ship);
      expect(ship.previousPort).toBe(previousPort);

File: __tests__/Port.test.js (hunk 9 above)
Comment (Chris-ONeill45) [reply __tests__/Port.test.js:37]: My code is what the track walkthrough I suggested i write. I have amended the functionality, do I still need to amend this test?

File: __tests__/Port.test.js (hunk 7 above)
Comment (PerryBaran) [reply __tests__/Port.test.js:23]: You should definitely have this test written as I have. If you are defining a value to be something specific you should make sure that it is that value

File: __tests__/Ship.test.js (hunk 12 above)
Comment (PerryBaran) [reply __tests__/Ship.test.js:59]: Ye that works

File: __tests__/Port.test.js (hunk 9 above)
Comment (PerryBaran) [reply __tests__/Port.test.js:37]: After some thought I think it would be best to have both of these tests. The test I suggested could also give a false positive if you changed your original `removeShip` functionality from using `.pop` to `.shift`, however, with both tests at least one of them would fail if you implemented it wrong.
